core/src/main/java/hudson/model/Job.java

```diff
@@ -346,12 +347,49 @@ public boolean isKeepDependencies() {
     }
 
     /**
-     * Allocates a new buildCommand number.
+     * Allocates a new build number.
+     * @see BuildNumberAssigner
      */
-    public synchronized int assignBuildNumber() throws IOException {
-        int r = nextBuildNumber++;
-        saveNextBuildNumber();
-        return r;
+    public int assignBuildNumber() throws IOException {
+        var bna = ExtensionList.lookupFirst(BuildNumberAssigner.class).assignBuildNumber(this);
+        if (bna.needsSave) {
+            saveNextBuildNumber();
+        }
+        return bna.assignedNumber;
+    }
+
+    /**
+     * Alternate strategy for assigning build numbers.
+     * @see #assignBuildNumber()
+     */
+    @Restricted(Beta.class)
+    public interface BuildNumberAssigner extends ExtensionPoint {
+        /**
+         * Strategy outcome.
+         * @param assignedNumber the return value for {@link #assignBuildNumber()}
+         * @param needsSave whether {@link #saveNextBuildNumber} should be called afterwards
+         */
+        record BuildNumberAssignment(int assignedNumber, boolean needsSave) {};
+
+        /**
+         * Assign a unique number for a new build of a given job.
+         * @see #getNextBuildNumber
+         * @see #fastUpdateNextBuildNumber
+         */
+        BuildNumberAssignment assignBuildNumber(Job<?, ?> job) throws IOException;
```

Comment:
The only alternative I can think of would be, but it looks as awkward.
```
@FunctionalInterface
public interface Saver<T> {
   void save(T t) throws IOException;
}
```
and
```
BuildNumberAssignment assignBuildNumber(Job<?, ?> job, Saver<Job<?,?> saver) throws IOException;
```

Reply:
I guess you mean

```java
int assignBuildNumber(Job<?, ?> job, Saver<Job<?,?>> saver) throws IOException;
```

or more simply (since we only want to allow _this_ job to be saved)

```java
interface Saver {
    void saveNextBuildNumber() throws IOException;
}
int assignBuildNumber(Job<?, ?> job, Saver saver) throws IOException;
```

called as

```java
return assignBuildNumber(this, this::saveNextBuildNumber);
```

Would work. I am not sure if it is any clearer.

Reply:
Yes. Not vastly clearer but at least the implementation has full control on when save occurs.